BUG: IntervalIndex is_non_overlapping_monotonic is incorrect when endpoints are closed on both sides

alysivji opened the pull request:
- [x] closes #16560
 - [x] tests added / passed
 - [x] passes ``git diff upstream/master --name-only -- '*.py' | flake8 --diff``
 - [x] whatsnew entry

alysivji committed:
Fix IntervalIndex is_non_overlapping_monotonic bug where closed='both' returned incorrect values

jreback commented on pandas/core/indexes/interval.py:
```diff
@@ -556,8 +556,12 @@ def is_non_overlapping_monotonic(self):
+        if self.closed == 'both':
+            return ((self.right[:-1] < self.left[1:]).all() or
+                    (self.left[:-1] > self.right[1:]).all())
+        else:
```
don't use the else here

add a comment on these cases (left/right).

neighter case? (short circuit to False)

alysivji replied:
Working on this now. Why would the neither case short circuit to False? Can't we have intervals [(0, 1), (1, 2)] where it is non-overlapping and also cases where it is overlapping [(0, 1), (.5, 3)]

Thanks!

jreback replied:
this is checking monotonicity as well. so you can do the same for neither.

jreback commented on pandas/tests/indexes/test_interval.py:
```diff
@@ -371,10 +371,19 @@ def slice_locs_cases(self, breaks):
+    def test_is_non_overlapping_monotonic(self):
+        index = IntervalIndex.from_tuples(
+            [(0, 1), (2, 3), (4, 5)], closed='both')
+        assert index.is_non_overlapping_monotonic == True
```
use is True

jreback commented on pandas/tests/indexes/test_interval.py:
```diff
@@ -371,10 +371,19 @@ def slice_locs_cases(self, breaks):
+        assert index.is_non_overlapping_monotonic == True
+
+        index = IntervalIndex.from_breaks(range(4), closed='both')
+        assert index.is_non_overlapping_monotonic == False
```
use is False

jreback reviewed (changes requested):
some changes. add a whatsnew note. 0.21.0 conversion bug fixes

alysivji committed:
Clean up code / add comments + additional test cases. Update WhatsNew with change information.

alysivji commented:
Fixed the issues from the review, except for changing ```==``` to ```is``` in the asserts. Using ```is``` resulted in assert errors so I kept it the way it was originally.

Added additional tests and updated the whatsnew section in the docs. Thanks.

jschendel commented:
ping @jreback : This can be closed, as the issue was resolved by #17238

gfyoung commented:
@jschendel : Confirmed and closing.
